webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueWriter.scala

```diff
@@ -593,13 +593,97 @@ class ClassicTechniqueWriter(basePath : String, parameterTypeService: ParameterT
     s"""_method_reporting_context_v4("${component}", "${value}","${methodCall.id}")"""
   }
 
+  def reportingContextInBundle(args: Seq[String]) = {
+    s"_method_reporting_context_v4(${convertArgsToBundleCall(args)})"
+  }
+
+
+  def convertArgsToBundleCall(args:Seq[String]) : String = {
+    args.map(escapeCFEngineString(_)).map(""""${""" + _ + """}"""").mkString(",")
+  }
 
 
   def writeAgentFiles( technique : EditorTechnique, methods : Map[BundleName, GenericMethod] )  : IOResult[Seq[String]] = {
 
+    // increment of the bundle number in the technique, used by createCallingBundle
+    var bundleIncrement = 0
+
     val bundleParams = if (technique.parameters.nonEmpty) technique.parameters.map(_.name.canonify).mkString("(",",",")") else ""
 
-    def bundleMethodCall( parentBlocks : List[MethodBlock])(method : MethodElem) : List[String] = {
+    // generate a bundle which encapsulate the method_reporting_context + the actual method call
+    // and the method to call this bundle
+    // Params are:
+    // * condition when to call this bundle
+    // * the methodCall itself from the technique editor
+    // * the class parameter _value_
+    // * all the parameters already converted to cfengine format
+    // * if it's for the NAReports bundle (reverse the condition from if to unless)
+    // Returns the bundle agent, and the "promised_" usebundle => bundle_created
+    def createCallingBundle(condition          : String
+                          , call               : MethodCall
+                          , classParameterValue: String
+                          , params             : Seq[String]
+                          , forNaReport        : Boolean) = {
+      val promiser = call.id + "_${report_data.directive_id}"
+
+      val filterOnMethod = forNaReport match {
+        case false => "if"
+        case true  => "unless"
+      }
+
+      // Reporting argument
+      val reportingArgs = escapeCFEngineString(call.component) ::
+                          escapeCFEngineString(classParameterValue) ::
+                          call.id :: Nil
+
+      // create the bundle arguments:
+      // there are 3 arguments corresponding to the reportingArgs, (nned to be quoted)
+      // the rest is for the methodArgs.
+      val bundleArguments = reportingArgs.map( x  => s""""${x}"""")  ::: params.toList
+
+      val argsList = bundleArguments.zipWithIndex.map {case (_, id) => "arg_" + id}
+
+      val bundleName = (call.id + "_" + bundleIncrement).replaceAll("-", "_")
```

Comment:
this may be not unique. We ought to prefix it with the technique name probably
Most likely technique_name_generic_method_name_incr would be more suited

Reply:
what would make this not unique ? rudder_reporting ?

Reply:
we were having problems of non execution due to call.id being reused when the technique was applied twice, and the problem was that the promiser was not unique

Reply:
but the technique file will be created once